### ISSUE TITLE
winrm send_input failed AKA Error 500

### ISSUE THREAD

**Issue by AleksNovak** [ansible/ansible#27556]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
WinRM

##### ANSIBLE VERSION

```bash
$ ansible --version
ansible 2.3.1.0
  config file = ~/ansible-deploy/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
$ pip list
ansible (2.3.1.0)
asn1crypto (0.22.0)
bcrypt (3.1.3)
certifi (2017.7.27.1)
cffi (1.10.0)
chardet (3.0.4)
cryptography (2.0.2)
enum34 (1.1.6)
idna (2.5)
ipaddress (1.0.18)
Jinja2 (2.9.6)
MarkupSafe (1.0)
ntlm-auth (1.0.5)
ordereddict (1.1)
paramiko (2.2.1)
pip (9.0.1)
pyasn1 (0.3.1)
pycparser (2.18)
pycrypto (2.6.1)
pykerberos (1.1.14)
PyNaCl (1.1.2)
pywinrm (0.2.2)
PyYAML (3.12)
requests (2.18.2)
requests-kerberos (0.11.0)
requests-ntlm (1.0.0)
setuptools (36.2.6)
six (1.10.0)
urllib3 (1.22)
wheel (0.29.0)
xmltodict (0.11.0)
```

##### CONFIGURATION
```ini
[defaults]
callback_plugins = ansible/callback_plugins/
transport = ssh
remote_user = root
host_key_checking = False
remote_tmp = /tmp
roles_path = ./ansible/roles
log_path = ./log/ansible.log
inventory = ./ansible/inventory.ini


retry_files_enabled = True
retry_files_save_path = ~/.ansible-retry

ansible_managed = This file is managed by ansible, don't make changes here - they will be overwritten.
[ssh_connection]
scp_if_ssh = True
pipelining = True
```

##### OS / ENVIRONMENT
CentOS Linux release 7.3.1611 (Core)
Python VirtualEnv 2.7

##### SUMMARY
I have two windows servers:
 - WinBox2008
    Windows 2008 R2 SP1
    PowerShell 3.0
    DotNet 4.5
 - WinBox2012
    Windows 2012 R2
    PowerShell 4.0
    DotNet 4.5

I can successfully connect to WinBox2008.
I **cannot** successfully connect to WinBox2012.

##### STEPS TO REPRODUCE

```yaml
[Windows]
WinBox2008.enter.prise
WinBox2012.enter.prise

[Windows:vars]
ansible_user=username@enter.prise
#ansible_password=password
#ansible_port: 5986
ansible_connection=winrm
#ansible_winrm_realm: ENTER.PRISE
ansible_winrm_transport=kerberos
ansible_winrm_kerberos_delegation=yes
ansible_winrm_server_cert_validation=ignore
```

```bash
$ klist
Ticket cache: FILE:/tmp/krb5cc_1000
Default principal: username@ENTER.PRISE

Valid starting       Expires              Service principal
08/01/2017 10:49:22  08/01/2017 20:49:22  krbtgt/INTER.PRISE@ENTER.PRISE
	renew until 08/08/2017 10:49:18
08/01/2017 10:50:33  08/01/2017 20:49:22  HTTP/WinBox2008.enter.prise@ENTER.PRISE
	renew until 08/08/2017 10:49:18
08/01/2017 10:50:33  08/01/2017 20:49:22  HTTP/WinBox2012.enter.prise@ENTER.PRISE
	renew until 08/08/2017 10:49:18
```
```bash
$ nc -zv -w 5 WinBox2008 5986
Connection to si026 5986 port [tcp/*] succeeded!
$ nc -zv -w 5 WinBox2012 5986
Connection to si900 5986 port [tcp/*] succeeded!
```

```yaml
---
- name: Prepare config for windows
  hosts: Windows
  tasks:
    - name: Test
      debug:
        msg: "{{ inventory_hostname }}"
```

Here is the ansible output: https://gist.github.com/AleksNovak/7b4b690a972806f49e7c5e63f016278a

##### EXPECTED RESULTS
Connect to both servers

##### ACTUAL RESULTS

```bash
 [WARNING]: FATAL ERROR DURING FILE TRANSFER: Traceback (most recent call last):
 File "~/python27/lib/python2.7/site-
 packages/ansible/plugins/connection/winrm.py", line 267, in _winrm_exec
 self._winrm_send_input(self.protocol, self.shell_id, command_id, data, eof=is_last)   File "~/python27/lib/python2.7/site-packages/ansible/plugins/connection/winrm.py",
 line 248, in _winrm_send_input protocol.send_message(xmltodict.unparse(rq))   File "~/python27/lib/python2.7/site-packages/winrm/protocol.py", line 207, in send_message     return self.transport.send_message(message)   File "~/python27/lib/python2.7/site-packages/winrm transport.py", line 202, in send_message     raise WinRMTransportError('http', error_message) WinRMTransportError: (u'http', u'Bad HTTP response returned from server. Code 500')

 <WinBox2012.ENTER.PRISE> WINRM CLOSE SHELL:
 90A5E767-35B3-4679-9AF1-C043555B4C66
 fatal: [WinBox2012.ENTER.PRISE]: FAILED! => {
     "failed": true
     }

     MSG:

     winrm send_input failed

     META: ran handlers

```

Update: Some test I’ve made on my WinBox

WinBox2012:
```PowerShell
PS C:\Windows\system32> Get-ChildItem WSMan:\localhost\Service\Auth

   WSManConfig : Microsoft.WSMan.Management\WSMan::localhost\Service\Auth

Type            Name                           SourceOfValue   Value
----            ----                           -------------   -----
System.String   Basic                                          true
System.String   Kerberos                                       true
System.String   Negotiate                      GPO             true
System.String   Certificate                                    false
System.String   CredSSP                                        false
System.String   CbtHardeningLevel                              Relaxed

PS C:\Windows\system32> $httpsOptions = New-PSSessionOption -SkipCACheck -SkipCNCheck -SkipRevocationCheck
PS C:\Windows\system32> New-PSSession -UseSSL -ComputerName localhost -SessionOption $httpsOptions

 Id Name            ComputerName    State         ConfigurationName     Availability
 -- ----            ------------    -----         -----------------     ------------
  1 Session1        localhost       Opened        Microsoft.PowerShell     Available
```

WinBox2008
```PowerShell
PS C:\Windows\system32> Get-ChildItem WSMan:\localhost\Service\Auth

   WSManConfig : Microsoft.WSMan.Management\WSMan::localhost\Service\Auth

Type            Name                           SourceOfValue   Value                                          
----            ----                           -------------   -----                                          
System.String   Basic                                          true                                           
System.String   Kerberos                                       true                                           
System.String   Negotiate                      GPO             true                                           
System.String   Certificate                                    false                                          
System.String   CredSSP                                        false                                          
System.String   CbtHardeningLevel                              Relaxed                                        



PS C:\Windows\system32> $httpsOptions = New-PSSessionOption -SkipCACheck -SkipCNCheck -SkipRevocationCheck

PS C:\Windows\system32> New-PSSession -UseSSL -ComputerName WinBox2012 -SessionOption $httpsOptions

 Id Name            ComputerName    State         ConfigurationName     Availability
 -- ----            ------------    -----         -----------------     ------------
  2 Session2        WinBox2012      Opened        Microsoft.PowerShell     Available
```

Update: I just ran again `ConfigureRemotingForAnsible.ps1` on the impacted box. Still no improvement.
WinRM Config from WinBox2012
```PowerShell
PS C:\Windows\system32> winrm get winrm/config
Config
    MaxEnvelopeSizekb = 500
    MaxTimeoutms = 60000
    MaxBatchItems = 32000
    MaxProviderRequests = 4294967295
    Client
        NetworkDelayms = 5000
        URLPrefix = wsman
        AllowUnencrypted = false
        Auth
            Basic = true
            Digest = true
            Kerberos = true
            Negotiate = true
            Certificate = true
            CredSSP = false
        DefaultPorts
            HTTP = 5985
            HTTPS = 5986
        TrustedHosts
    Service
        RootSDDL = O:NSG:BAD:P(A;;GA;;;BA)(A;;GR;;;IU)S:P(AU;FA;GA;;;WD)(AU;SA;GXGW;;;WD)
        MaxConcurrentOperations = 4294967295
        MaxConcurrentOperationsPerUser = 1500
        EnumerationTimeoutms = 240000
        MaxConnections = 300
        MaxPacketRetrievalTimeSeconds = 120
        AllowUnencrypted = false
        Auth
            Basic = true
            Kerberos = true
            Negotiate = true [Source="GPO"]
            Certificate = false
            CredSSP = false
            CbtHardeningLevel = Relaxed
        DefaultPorts
            HTTP = 5985
            HTTPS = 5986
        IPv4Filter = * [Source="GPO"]
        IPv6Filter = * [Source="GPO"]
        EnableCompatibilityHttpListener = false
        EnableCompatibilityHttpsListener = false
        CertificateThumbprint
        AllowRemoteAccess = true [Source="GPO"]
    Winrs
        AllowRemoteShellAccess = true
        IdleTimeout = 7200000
        MaxConcurrentUsers = 10
        MaxShellRunTime = 2147483647
        MaxProcessesPerShell = 25
        MaxMemoryPerShellMB = 1024
        MaxShellsPerUser = 30
```

**Comment by jhawkesworth** [comment]
Suggestions.
Try longer winrm timeouts:
ansible_winrm_operation_timeout_sec: 60
ansible_winrm_read_timeout_sec: 70

Also check to see if your affected machines are running scheduled maintenance (this can happen frequently if you often restore windows machines from snapshots).  Maintenance can cause heavy CPU usage which can affect other parts of the system.

Please can you try with above increased timeout settings?

needs_info

**Comment by AleksNovak** [comment]
Hi, 
This issue is happening all the time. Until now, I haven’t been able to connect to that box.
There is nothing on that server that can cause heavy load.

PS: I’m adding more powershell test on the issue.

I’ve added the requested options:

```ini
[Windows]
WinBox2012.enter.prise

[Windows:vars]
ansible_user=username@ENTER.PRISE
ansible_password=password
#ansible_port: 5986
ansible_connection=winrm
#ansible_winrm_realm: ENTER.PRISE
ansible_winrm_transport=kerberos
ansible_winrm_kerberos_delegation=yes
ansible_winrm_server_cert_validation=ignore
ansible_winrm_operation_timeout_sec=60
ansible_winrm_read_timeout_sec=70
```

```bash
TASK [Gathering Facts] ******************************************************************************************************************************
<WinBox2012.ENTER.PRISE> ESTABLISH WINRM CONNECTION FOR USER: username@ENTER.PRISE on PORT 5986 TO WinBox2012.ENTER.PRISE
creating Kerberos CC at /tmp/tmpK4B4tB
calling kinit for principal username@ENTER.PRISE
kinit succeeded for principal username@ENTER.PRISE 
<WinBox2012.enter.prise> WINRM CONNECT: transport=kerberos endpoint=https://WinBox2012.enter.prise:5986/wsman
<WinBox2012.enter.prise> WINRM OPEN SHELL: CDE099A2-0EA0-4734-BFE7-F1921E960962
EXEC (via pipeline wrapper)
<WinBox2012.enter.prise> WINRM EXEC 'PowerShell' ['-NoProfile', '-NonInteractive', '-ExecutionPolicy', 'Unrestricted', '-EncodedCommand', 'IwBSAGUAcQB1AGkAcgBlAHMAIAAtAFYAZQByAHMAaQBvAG4AIAAzAC4AMAAKAGIAZQBnAGkAbgAgAHsACgAkAEQAZQBiAHUAZwBQAHIAZQBmAGUAcgBlAG4AYwBlACAAPQAgACIAQwBvAG4AdABpAG4AdQBlACIACgAkAEUAcgByAG8AcgBBAGMAdABpAG8AbgBQAHIAZQBmAGUAcgBlAG4AYwBlACAAPQAgACIAUwB0AG8AcAAiAAoAUwBlAHQALQBTAHQAcgBpAGMAdABNAG8AZABlACAALQBWAGUAcgBzAGkAbwBuACAAMgAKAGYAdQBuAGMAdABpAG8AbgAgAEMAbwBuAHYAZQByAHQAVABvAC0ASABhAHMAaAB0AGEAYgBsAGUARgByAG8AbQBQAHMAQwB1AHMAdABvAG0ATwBiAGoAZQBjAHQAIAAoACQAbQB5AFAAcwBPAGIAagBlAGMAdAApAHsACgAkAG8AdQB0AHAAdQB0ACAAPQAgAEAAewB9ADsACgAkAG0AeQBQAHMATwBiAGoAZQBjAHQAIAB8ACAARwBlAHQALQBNAGUAbQBiAGUAcgAgAC0ATQBlAG0AYgBlAHIAVAB5AHAAZQAgACoAUAByAG8AcABlAHIAdAB5ACAAfAAgACUAIAB7AAoAJAB2AGEAbAAgAD0AIAAkAG0AeQBQAHMATwBiAGoAZQBjAHQALgAoACQAXwAuAG4AYQBtAGUAKQA7AAoASQBmACAAKAAkAHYAYQBsACAALQBpAHMAIABbAHAAcwBvAGIAagBlAGMAdABdACkAIAB7AAoAJAB2AGEAbAAgAD0AIABDAG8AbgB2AGUAcgB0AFQAbwAtAEgAYQBzAGgAdABhAGIAbABlAEYAcgBvAG0AUABzAEMAdQBzAHQAbwBtAE8AYgBqAGUAYwB0ACAAJAB2AGEAbAAKAH0ACgAkAG8AdQB0AHAAdQB0AC4AKAAkAF8ALgBuAGEAbQBlACkAIAA9ACAAJAB2AGEAbAAKAH0ACgByAGUAdAB1AHIAbgAgACQAbwB1AHQAcAB1AHQAOwAKAH0ACgAjACAAcwB0AHIAZQBhAG0AIABKAFMATwBOACAAaQBuAGMAbAB1AGQAaQBuAGcAIABiAGUAYwBvAG0AZQBfAHAAdwAsACAAcABzAF8AbQBvAGQAdQBsAGUAXwBwAGEAeQBsAG8AYQBkACwAIABiAGkAbgBfAG0AbwBkAHUAbABlAF8AcABhAHkAbABvAGEAZAAsACAAYgBlAGMAbwBtAGUAXwBwAGEAeQBsAG8AYQBkACwAIAB3AHIAaQB0AGUAXwBwAGEAeQBsAG8AYQBkAF8AcABhAHQAaAAsACAAcAByAGUAcwBlAHIAdgBlACAAZABpAHIAZQBjAHQAaQB2AGUAcwAKACMAIABlAHgAZQBjACAAcgB1AG4AcwBwAGEAYwBlACwAIABjAGEAcAB0AHUAcgBlACAAbwB1AHQAcAB1AHQALAAgAGMAbABlAGEAbgB1AHAALAAgAHIAZQB0AHUAcgBuACAAbQBvAGQAdQBsAGUAIABvAHUAdABwAHUAdAAKACQAagBzAG8AbgBfAHIAYQB3ACAAPQAgACIAIgAKAH0ACgBwAHIAbwBjAGUAcwBzACAAewAKACQAaQBuAHAAdQB0AF8AYQBzAF8AcwB0AHIAaQBuAGcAIAA9ACAAWwBzAHQAcgBpAG4AZwBdACQAaQBuAHAAdQB0AAoAJABqAHMAbwBuAF8AcgBhAHcAIAArAD0AIAAkAGkAbgBwAHUAdABfAGEAcwBfAHMAdAByAGkAbgBnAAoAfQAKAGUAbgBkACAAewAKAEkAZgAgACgALQBuAG8AdAAgACQAagBzAG8AbgBfAHIAYQB3ACkAIAB7AAoAVwByAGkAdABlAC0ARQByAHIAbwByACAAIgBuAG8AIABpAG4AcAB1AHQAIABnAGkAdgBlAG4AIgAgAC0AQwBhAHQAZQBnAG8AcgB5ACAASQBuAHYAYQBsAGkAZABBAHIAZwB1AG0AZQBuAHQACgB9AAoAJABwAGEAeQBsAG8AYQBkACAAPQAgAEMAbwBuAHYAZQByAHQAVABvAC0ASABhAHMAaAB0AGEAYgBsAGUARgByAG8AbQBQAHMAQwB1AHMAdABvAG0ATwBiAGoAZQBjAHQAIAAoAEMAbwBuAHYAZQByAHQARgByAG8AbQAtAEoAcwBvAG4AIAAkAGoAcwBvAG4AXwByAGEAdwApAAoAIwAgAFQATwBEAE8AOgAgAGgAYQBuAGQAbABlACAAYgBpAG4AYQByAHkAIABtAG8AZAB1AGwAZQBzAAoAIwAgAFQATwBEAE8AOgAgAGgAYQBuAGQAbABlACAAcABlAHIAcwBpAHMAdABlAG4AYwBlAAoAJABhAGMAdABpAG8AbgBzACAAPQAgACQAcABhAHkAbABvAGEAZAAuAGEAYwB0AGkAbwBuAHMACgAjACAAcABvAHAAIAAwAHQAaAAgAGEAYwB0AGkAbwBuACAAYQBzACAAZQBuAHQAcgB5AHAAbwBpAG4AdAAKACQAZQBuAHQAcgB5AHAAbwBpAG4AdAAgAD0AIAAkAHAAYQB5AGwAbwBhAGQALgAoACQAYQBjAHQAaQBvAG4AcwBbADAAXQApAAoAJABwAGEAeQBsAG8AYQBkAC4AYQBjAHQAaQBvAG4AcwAgAD0AIAAkAHAAYQB5AGwAbwBhAGQALgBhAGMAdABpAG8AbgBzAFsAMQAuAC4AOQA5AF0ACgAkAGUAbgB0AHIAeQBwAG8AaQBuAHQAIAA9ACAAWwBTAHkAcwB0AGUAbQAuAFQAZQB4AHQALgBFAG4AYwBvAGQAaQBuAGcAXQA6ADoAVQBUAEYAOAAuAEcAZQB0AFMAdAByAGkAbgBnACgAWwBTAHkAcwB0AGUAbQAuAEMAbwBuAHYAZQByAHQAXQA6ADoARgByAG8AbQBCAGEAcwBlADYANABTAHQAcgBpAG4AZwAoACQAZQBuAHQAcgB5AHAAbwBpAG4AdAApACkACgAjACAAbABvAGEAZAAgAHQAaABlACAAYwB1AHIAcgBlAG4AdAAgAGEAYwB0AGkAbwBuACAAZQBuAHQAcgB5AHAAbwBpAG4AdAAgAGEAcwAgAGEAIABtAG8AZAB1AGwAZQAgAGMAdQBzAHQAbwBtACAAbwBiAGoAZQBjAHQAIAB3AGkAdABoACAAYQAgAFIAdQBuACAAbQBlAHQAaABvAGQACgAkAGUAbgB0AHIAeQBwAG8AaQBuAHQAIAA9ACAATgBlAHcALQBNAG8AZAB1AGwAZQAgAC0AUwBjAHIAaQBwAHQAQgBsAG8AYwBrACAAKABbAHMAYwByAGkAcAB0AGIAbABvAGMAawBdADoAOgBDAHIAZQBhAHQAZQAoACQAZQBuAHQAcgB5AHAAbwBpAG4AdAApACkAIAAtAEEAcwBDAHUAcwB0AG8AbQBPAGIAagBlAGMAdAAKAFMAZQB0AC0AVgBhAHIAaQBhAGIAbABlACAALQBTAGMAbwBwAGUAIABnAGwAbwBiAGEAbAAgAC0ATgBhAG0AZQAgAGMAbwBtAHAAbABlAHgAXwBhAHIAZwBzACAALQBWAGEAbAB1AGUAIAAkAHAAYQB5AGwAbwBhAGQAWwAiAG0AbwBkAHUAbABlAF8AYQByAGcAcwAiAF0AIAB8ACAATwB1AHQALQBOAHUAbABsAAoAIwAgAGQAeQBuAGEAbQBpAGMAYQBsAGwAeQAgAGMAcgBlAGEAdABlAC8AbABvAGEAZAAgAG0AbwBkAHUAbABlAHMACgBGAG8AcgBFAGEAYwBoACAAKAAkAG0AbwBkACAAaQBuACAAJABwAGEAeQBsAG8AYQBkAC4AcABvAHcAZQByAHMAaABlAGwAbABfAG0AbwBkAHUAbABlAHMALgBHAGUAdABFAG4AdQBtAGUAcgBhAHQAbwByACgAKQApACAAewAKACQAZABlAGMAbwBkAGUAZABfAG0AbwBkAHUAbABlACAAPQAgAFsAUwB5AHMAdABlAG0ALgBUAGUAeAB0AC4ARQBuAGMAbwBkAGkAbgBnAF0AOgA6AFUAVABGADgALgBHAGUAdABTAHQAcgBpAG4AZwAoAFsAUwB5AHMAdABlAG0ALgBDAG8AbgB2AGUAcgB0AF0AOgA6AEYAcgBvAG0AQgBhAHMAZQA2ADQAUwB0AHIAaQBuAGcAKAAkAG0AbwBkAC4AVgBhAGwAdQBlACkAKQAKAE4AZQB3AC0ATQBvAGQAdQBsAGUAIAAtAFMAYwByAGkAcAB0AEIAbABvAGMAawAgACgAWwBzAGMAcgBpAHAAdABiAGwAbwBjAGsAXQA6ADoAQwByAGUAYQB0AGUAKAAkAGQAZQBjAG8AZABlAGQAXwBtAG8AZAB1AGwAZQApACkAIAAtAE4AYQBtAGUAIAAkAG0AbwBkAC4ASwBlAHkAIAB8ACAASQBtAHAAbwByAHQALQBNAG8AZAB1AGwAZQAgAC0AVwBhAHIAbgBpAG4AZwBBAGMAdABpAG8AbgAgAFMAaQBsAGUAbgB0AGwAeQBDAG8AbgB0AGkAbgB1AGUAIAB8ACAATwB1AHQALQBOAHUAbABsAAoAfQAKACQAbwB1AHQAcAB1AHQAIAA9ACAAJABlAG4AdAByAHkAcABvAGkAbgB0AC4AUgB1AG4AKAAkAHAAYQB5AGwAbwBhAGQAKQAKAFcAcgBpAHQAZQAtAE8AdQB0AHAAdQB0ACAAJABvAHUAdABwAHUAdAAKAH0A']
 [WARNING]: FATAL ERROR DURING FILE TRANSFER: Traceback (most recent call last):   File "~/python27/lib/python2.7/site-packages/ansible/plugins/connection/winrm.py", line 267, in winrm_exec self._winrm_send_input(self.protocol, self.shell_id, command_id, data, eof=is_last) File "~/python27/lib/python2.7/site-packages/ansible/plugins/connection/winrm.py", line 248, in _winrm_send_input protocol.send_message(xmltodict.unparse(rq))   File "~/python27/lib/python2.7/site-packages/winrm/protocol.py", line 207, in send_message     return self.transport.send_message(message)   File "~/python27/lib/python2.7/site-packages/winrm/transport.py", line 202, in send_message     raise WinRMTransportError('http', error_message) WinRMTransportError: (u'http', u'Bad HTTP response returned from server. Code 500')

<WinBox2012.enter.prise> WINRM CLOSE SHELL: CDE099A2-0EA0-4734-BFE7-F1921E960962
fatal: [WinBox2012.enter.prise]: FAILED! => {
    "failed": true
}

MSG:
winrm send_input failed

```

**Comment by jhawkesworth** [comment]
Is there any information in the event log of WinBox2012.ENTER.PRISE at the time the login is attempted by ansible?

needs_info

**Comment by AleksNovak** [comment]
It’s really hard for me to debug that part as I’m completely lost in Windows log system.
What should I look for ?

**Comment by AleksNovak** [comment]
So, I did a new connection test and I found a way to extract EventLog from the WinBox2012.
ps: sorry for my french.

[analytic.txt](https://github.com/ansible/ansible/files/1191157/analytic.txt)
[operationnal.txt](https://github.com/ansible/ansible/files/1191158/operationnal.txt)

**Comment by jborean93** [comment]
@AleksNovak, there is no easy way to get logs from the WinRM service but I found this guide can be helpful if you wish to look more into the logs https://blogs.technet.microsoft.com/heyscriptingguy/2015/10/05/troubleshoot-winrm-with-powershellpart-1/ (looks like you have already managed it though :)). Unfortunately even the level of detail you can get from the debug WinRM logs is pretty abysmal for most purposes so I usually hit a brick wall here.

As for the issue itself, I believe it is because the SSL connection get's wedged and we have seen it happen before in other scenarios. We have found the combination of the OpenSSL version Python was compiled with and the cipher suites that Windows sets cause this issue.

You can use a program like IIS Crypto https://www.nartac.com/Products/IISCrypto to set the cipher suite priority order, enable/disable different protocols as well as set standard best practices. If you are going to play around I would recommend taking snapshots of your server as you can break RDP and other services reliant on SChannel if you play around too much.

One thing you can do though to totally rule out this SSL wedging issue is to run WinRM over HTTP for some tests. To do this you can either

* Install this version of pywinrm https://github.com/diyan/pywinrm/pull/156
* Change the port Ansible uses to be `5985`
* Add the vars; `ansible_winrm_transport=ntlm` and `ansible_winrm_message_encryption=auto`
* Test your playbooks

or

* Allow unencrypted messages by running `winrm set winrm/config/service @{AllowUnencrypted="true"}` on the Windows host
* Change the port Ansible uses to be `5985`
* Test your playbooks
* When you are finished run `winrm set winrm/config/service @{AllowUnencrypted="false"}`

If doing the latter it is quite important you don't move forward with allowing unencrypted messages as it is quite insecure, it is a useful tool for debugging and testing but shouldn't be used in a production scenario.

**Comment by jhawkesworth** [comment]
Trying a more recent python version might be a quicker way to get a different version of OpenSSL to take effect, if you are able to upgrade python.

**Comment by AleksNovak** [comment]
Hi,
I didn’t give you those information. But, I first tried to do it with a python 3.6 VEnv.
Also, the Windows server receive the WinRM certificate through GPO. So, they all have the same.

I tested other Windows Server 2012 with success. Only one have this weird issue.

**Comment by asovern-mitre** [comment]
I'm running Python 2.7.12 and have the same "winrm send_input" issue when using ansible 2.3.x . I tried rolling back to ansible version 2.2.3 and the issue went away & I was able to hit my host with winrm again. It seems like a code change with the way ansible interfaces with pywinrm changed from 2.2 to 2.3.
AleksNovak, have you tried testing with an older version of Ansible to see if the connection goes through?

**Comment by jhawkesworth** [comment]
@AleksNovak It really feels like it has to be a problem with the windows server itself.
I am not sure if python 3.x is fully working with ansible's windows support at the moment, but perhaps try a different 2.x version if you can.
The only other thing I can think of is perhaps there is an SPN configured in active directory for the WS2012 machine.  I have heard of users having problems connecting when there is an SPN in place.

Here's one link I found on the topic of SPN interferring with WinRM.
 https://dba.stackexchange.com/questions/129286/how-do-i-enable-both-powershell-remoting-and-spn-for-sql-server-reporting
I am not sure if the fix described will work for pywinrm, but worth asking Active Directory adminstrators if there is an SPN configured for the machine.

**Comment by jborean93** [comment]
Ansible 2.3 changed the way powershell modules were executable on the server which would be why you see an issue on a newer version. Before here is what Ansible did

* Create a new temp folder
* Copy across the module including the module utils to that temp folder
* Execute the temp module at that path

Now this is what happens

* Execute a powershell script in memory that accepts an input
* Send the module definition through stdin to that input
* When input is complete it will execute that code

This way no file is created on the host and the execution itself is a lot quicker. Unfortunately this seems to have exacerbated the SSL wedging issue or some other host configuration is causing a problem as it is a different operation to send input compared to just executing a command.

@AleksNovak were you able to solve this on your environment or are you still having problems?

**Comment by VitaliStanevich** [comment]
Having the same issu while trying to manage a VM based on Windows 8.1 from CentOS 7. 
I have Ansible 2.3.2.0. Also checked on Ansible 2.3.1 and 2.4 - same troubles.
Are there any updates solving the issue? 
@AleksNovak

**Comment by roded** [comment]
Following @asovern's suggestion we're sticking to version 2.2.3.0.
This issue is obviously blocking Windows shops from updating.
Can anyone update on the priority of this issue? Is a fix planned?

**Comment by jborean93** [comment]
I've had a further look into the logs that were given and found the following entry in the analytics logs that was sent through (note this is translated by Google translate from French to English)

```
Information, 01/08/2017 17: 40: 56, Microsoft-Windows-WinRM, 288, WSMan API Call, Plug-in indicating a completed operation for Send
Error, 01/08/2017 17: 40: 56, Microsoft-Windows-WinRM, 1840, Operation Winrm, "An error occurred while processing an operation.
Error code: 1460
Error String: <f: WSManFault xmlns: f = "" http://schemas.microsoft.com/wbem/wsman/1/wsmanfault "" Code = "" 1460 "" Machine = "" windows-host "" > <f: Message> <f: ProviderFault provider = "" Shell cmd plugin "" path = ""% systemroot% \ system32 \ winrscmd.dll ""> This operation was completed because the timeout has expired. </ F: ProviderFault> </ f: Message> </ f: WSManFault> "
```

This means that the WinRM operation failed with error code 1460 and sent back the error (HTTP status 500). This error is caused by a timeout when processing the values, as for why it is timing out I'm not sure without seeing more info.

For others looking at this issue, here is some background information that might explain more as to what is happening. Before Ansible 2.3, the winrm connection plugin worked by copying across the module file and executing it manually through the command line. From 2.3 to now, Ansible sends the module to execute over the stdin pipe so it never touches the disk. This was done for the following reasons

* The execution time was a lot quicker as there were less WinRM calls involved
* Allowed us to easily add in features like become and async properly as a generic wrapper framework was added in 2.3
* Didn't need to mess around with permissions or finding temporary directories as nothing touched the disk

From this change, Ansible now has a function to send data through to the stdin pipe created where it didn't before. It is from this function that the `send_input failed` error message has come from. There are no plans to go back to the pre 2.3 behaviour as we have not found a genuine reason why this would fail on a supported setup. Every time this issue is raised it is due to the host itself (or WinRM service) being broken in some way that is unrelated to Ansible.

The times where Ansible will raise this particular exception is when it receives an exception while sending data over the stdin pipe to the Windows host. This can mean any error that is returned from the WinRM service but is usually cased by a HTTP 500 error. Here are some things that I have found to cause this error;

* The SSL connection gets "wedged" and fails, I have explored this issue further above and have posted some workaround
* Someone is trying to execute a Python module on Windows
* The powershell executable is not on the `PATH` machine environment variable
* The WinRM service is bounced for some reason (could be disabled during a program install or a critical error is found)
* There is a lot of load on the server causing the WinRM process to slow down and eventually cause a timeout as it couldn't finish during the pre-negotiated timeout window

When Ansible receives a HTTP 500 code back from it's request, it means that the WinRM service encountered an internal error that it couldn't handle and decided to close the shell. While there is a tiny chance it could be the way Ansible sends data over to the host I have personally never encountered an issue over all the supported OS types unless I did something to cause on the things listed above.

It is unfortunate that this error message is so generic and does not include any more info but that is a problem with the underlying pywinrm library. When pywinrm receives a non 200 HTTP code it raises an exception but it doesn't provide the message that was returned from WinRM making it hard for Ansible to give a proper message. If pywinrm was modified to return this information we could make the error a bit more descriptive.

I am going to close this issue as the comments are moving away from what is the issue and I don't want one long post for something that can be caused by different issues.

If you are receiving this error, you can turn on the analytics and operational logging for WinRM to find out more info. I found this guide is good at explaining how to do this http://www.itprotoday.com/management-mobility/tools-troubleshooting-powershell-remoting-and-winrm-part-2. If you have this info and still believe it is a genuine bug related to Ansible itself and how it talks over WinRM then feel free to raise a bug and provide this information.

If it seems to be a general Windows error or just need help deciphering the event logs, please post the info on the IRC channel or mailing list instead so we can keep Github just for actual issues. You can access either of these 2 options here

   * IRC: #ansible on irc.freenode.net   
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

I know some of you feel that me closing this issue is wrong, but this page has become a generic catchall for people encountering this error which in this case can mean a wide range of issues.

If you are still having this issue please open a new Github issue and supply as much information as you can. For the `send_input failed` error, the error logs will pretty much be required for us to further investigate and without them your issue may be closed as well so please add them to your github issue.

**Comment by ghost** [comment]
This may help someone else facing the same issue. Among the list of things that may cause this error, I would also add:

- Older version of Powershell (e.g. 2.0) installed in the system (together with a newer version, e.g. 4.0)

Here are the steps I used to debug the issue:

- Run Powershell from the Windows start menu. Everything should work fine
- Launch cmd.exe and run "powershell.exe" . In my case, the command would fail with a weird error (this is what would cause the "HTTP 500" in Ansible)
- From the same command prompt, run "where powershell.exe" to verify if multiple instances of powershell.exe are installed
- Make sure to remove older version of Powershell (Remove Windows feature)
- If the old binary file is still there, delete it (in my case it was under C:\Windows)

This fixed connectivity between Ansible (2.4+) and a Windows Server 2012 R2.
